### ISSUE TITLE
Type Narrowing for hasAttribute and getAttribute

### ISSUE THREAD

**Issue by TheColorRed** [Microsoft/TypeScript#22238]
**TypeScript Version:** 2.7.1 and 2.8.0-insiders.20180127

Type guard should catch `hasAttribute` so that `getAttribute` doesn't return null.

In the following example, TS still says `getAttribute` could `return null`, however we previously checked for the attribute on the previous line.

```ts
let element = document.querySelector('#element')

if(element.hasAttribute('data-title')) {
  let el = element.getAttribute('data-title') // still returns "string | null"
}
```

**Expected Result:**

`element.getAttribute('data-title')` Return type should be `string`

**Actual Result:**

`element.getAttribute('data-title')` Return type is `string | null`

**Comment by weswigham** [comment]
I believe this would be an (very complex, but possible) override of `hasAttribute` and `getAttribute`  `Element` signatures over at the lib generator - [Microsoft/TSJS-lib-generator](https://github.com/Microsoft/TSJS-lib-generator).

**Comment by mhegazy** [comment]
Duplicate of https://github.com/Microsoft/TypeScript/issues/18781 and https://github.com/Microsoft/TypeScript/issues/13086. I do nto think we want to add such overloads to the library, but you are welcomed to add them.  @DanielRosenwasser has a [sample](https://github.com/Microsoft/TypeScript/issues/13086#issuecomment-269695308) how

**Comment by TheColorRed** [comment]
@mhegazy Isn't that for `Set` and `Map`? Isn't that different than `hasAttribute` and `getAttribute`?

**Comment by weswigham** [comment]
It's the same kinda situation; the types would look identical.
